File: ShapeByte/features/workout/item/data/ItemSet.swift
//
//  ItemSet.swift
//  ShapeByte
//
//  Created by Lang, Stefan [Shape Byte Tech] on 28.07.24.
//

import Foundation

protocol ItemProviding {
    var item: any Item { get }
}

/**
 Representing a single performance of either an excerise (push up, squat) or break or countdown
 */

enum ItemSet: Equatable, ItemProviding {
    case timed(item: any Item, duration: TimeInterval)

    var item: any Item {
        switch self {

        case .timed(let item, _):
            return item
        }
    }

    var duration: TimeInterval? {
        switch self {
        case .timed(_, let duration):
            return duration
        }
    }

    var isTimed: Bool {
        switch self {
        case .timed:
            return true
        }
    }

    // TODO: Test!
    static func == (lhs: ItemSet, rhs: ItemSet) -> Bool {
        switch (lhs, rhs) {
        case let (.timed(lhsItem, lhsDuration), .timed(rhsItem, rhsDuration)):
            return lhsItem.isEqualTo(rhsItem) && lhsDuration == rhsDuration
        }
    }
}

File: features/workout/data/contract/src/commonMain/kotlin/de/stefan/lang/shapebyte/features/workout/data/contract/item/ItemSet.kt
package de.stefan.lang.shapebyte.features.workout.data.contract.item

import kotlin.time.Duration
import kotlin.time.Duration.Companion.milliseconds
import kotlin.time.Duration.Companion.seconds

/**
Representing a single performance of either an exercise (push up, squat) or break or countdown
 */

public const val SECOND_IN_MILLISECONDS: Int = 1000

public sealed interface ItemSet {
    public sealed interface Timed : ItemSet {
        public val milliSecsRaw: Int

        public val seconds: Duration get() = (milliSecsRaw / SECOND_IN_MILLISECONDS).seconds
        public val milliseconds: Duration get() = milliSecsRaw.milliseconds

        public data class Seconds(public val durationSeconds: Int) : Timed {
            override val milliSecsRaw: Int = durationSeconds * SECOND_IN_MILLISECONDS
        }

        public data class Milliseconds(override val milliSecsRaw: Int) : Timed
    }

    public data class Repetition(public val repetitions: UInt? = null) : ItemSet
}
